src/client/components/form-editor/index.js

```diff
@@ -140,8 +151,84 @@ class FormEditor extends DataComponent {
         let handleElCreation = el => Promise.try( () => synch(el) ).then( () => create(el) );
         let els = [ intn, ...ppts ];
 
-        return Promise.all( els.map(handleElCreation) ).then( () => {
-          return Promise.all( ppts.map(add) ).then( () => add(intn) );
+        let runLayout = cy => {
+          let layout = cy.layout( _.assign( {}, getCyLayoutOpts(), {
+            animate: false,
+            randomize: true
+          } ) );
+
+          // if there is an existing layout disrupt it
+          if ( this.lastLayout ) {
+            this.layoutDisrupted = true;
+            this.lastLayout.stop();
+          }
+
+          this.lastLayout = layout;
+
+          let layoutDone = layout.promiseOn('layoutstop');
+
+          layout.run();
+
+          return layoutDone;
+        };
+
+        let updatePos = el => {
+          let cyEle = cy.getElementById( el.id() );
+          let pos =  _.clone( cyEle.position() );
+
+          el.reposition( pos );
+        };
+
+        let cy = new Cytoscape({
+          headless: true,
+          elements: makeCyEles( doc.elements() ),
+          layout: { name: 'preset' },
+          styleEnabled: true
+        });
+
+        // already existing elements are supposed to be locked during the layout
+        cy.elements().lock();
+
+        this.elesToAdd.push( ...els );
+
+        // create cy eles for the new elements that are not represented in cy
+        // yet because they were not in document while cy is created
+        let missingEles = this.elesToAdd.filter( el => cy.getElementById( el.id() ).length === 0 );
+        cy.add( makeCyEles( missingEles ) );
+
+        let layoutDone = runLayout( cy );
+
+        return layoutDone.then( () => {
+          // skip disrupted layouts
+          if ( this.layoutDisrupted ) {
+            this.layoutDisrupted = false;
+            return;
+          }
+
+          let isIntn = el => el.isInteraction();
+          let isPpt = el => el.isEntity();
+
+          // skip elements that are already being added
+          let elesToAdd = this.elesToAdd.filter( ( el ) => !this.beingAdded.has( el ) );
+          let pptsToAdd = elesToAdd.filter( isPpt );
+          let intnsToAdd = elesToAdd.filter( isIntn );
+
+          return Promise.all( elesToAdd.map( updatePos ) ).then( () => {
+            this.lastLayout = null;
+            els.forEach( el => this.beingAdded.add( el ) );
+
+            let postAdd = () => {
+              let toAddSet = new Set( elesToAdd );
```

Comment:
The main benefit of creating the set in this case would be performance, but the set is so small it's probably best to just use `elesToAdd.some()` in the `_.remove()` line to keep things simple and easy to read.

Reply:
I think in that case ``some()`` will require a function to check so would not it be simpler to directly using [``_.includes()``](https://lodash.com/docs/4.17.10#includes)?